app/src/full/java/io/homeassistant/companion/android/sensors/HealthConnectSensorManager.kt

```diff
@@ -0,0 +1,168 @@
+import androidx.health.connect.client.request.ReadRecordsRequest
+import androidx.health.connect.client.time.TimeRangeFilter
+import io.homeassistant.companion.android.common.R as commonR
+import io.homeassistant.companion.android.common.sensors.SensorManager
+import java.time.Instant
+import java.time.LocalDate
+import java.time.LocalDateTime
+import java.time.LocalTime
+import java.time.ZoneOffset
+import java.time.temporal.ChronoUnit
+import kotlinx.coroutines.runBlocking
+
+class HealthConnectSensorManager : SensorManager {
+    companion object {
+        val activeCaloriesBurned = SensorManager.BasicSensor(
+            id = "health_connect_active_calories_burned",
+            type = "sensor",
+            commonR.string.basic_sensor_name_active_calories_burned,
+            commonR.string.sensor_description_active_calories_burned,
+            "mdi:fire",
+            unitOfMeasurement = "kcal"
+        )
+
+        val totalCaloriesBurned = SensorManager.BasicSensor(
+            id = "health_connect_total_calories_burned",
+            type = "sensor",
+            commonR.string.basic_sensor_name_total_calories_burned,
+            commonR.string.sensor_description_total_calories_burned,
+            "mdi:fire",
+            unitOfMeasurement = "kcal"
+        )
+
+        val weight = SensorManager.BasicSensor(
+            id = "health_connect_weight",
+            type = "sensor",
+            commonR.string.basic_sensor_name_weight,
+            commonR.string.sensor_description_weight,
+            "mdi:weight",
+            unitOfMeasurement = "g"
+
+        )
+    }
+
+    override val name: Int
+        get() = commonR.string.sensor_name_health_connect
+
+    override fun requiredPermissions(sensorId: String): Array<String> {
+        return when {
+            (sensorId == activeCaloriesBurned.id) -> arrayOf(HealthPermission.getReadPermission(ActiveCaloriesBurnedRecord::class))
+            (sensorId == totalCaloriesBurned.id) -> arrayOf(HealthPermission.getReadPermission(TotalCaloriesBurnedRecord::class))
+            (sensorId == weight.id) -> arrayOf(HealthPermission.getReadPermission(WeightRecord::class))
+            else -> arrayOf()
+        }
+    }
+
+    override fun requestSensorUpdate(context: Context) {
+        val healthConnectClient: HealthConnectClient = HealthConnectClient.getOrCreate(context)
+
+        if (isEnabled(context, weight)) {
+            updateWeightSensor(context, healthConnectClient)
+        }
+        if (isEnabled(context, activeCaloriesBurned)) {
+            updateActiveCaloriesBurnedSensor(context, healthConnectClient)
+        }
+        if (isEnabled(context, totalCaloriesBurned)) {
+            updateTotalCaloriesBurnedSensor(context, healthConnectClient)
+        }
+    }
+
+    private fun updateTotalCaloriesBurnedSensor(context: Context, healthConnectClient: HealthConnectClient) {
+        val totalCaloriesBurnedRequest = runBlocking {
```

Comment:
are we able to avoid `runBlocking` everywhere I think we may be hitting another case pointed out in https://github.com/home-assistant/android/pull/4340 here? All sensor updates are already in a coroutine.

Reply:
yea given we need to get Google's approval, we should def minimize the amount of work in this PR